contracts/v0.6/swap/ERC20TokenPeg.sol

```diff
@@ -0,0 +1,116 @@
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import "./ISwap.sol";
+
+/**
+ * @title ERC20TokenPeg
+ * @notice Using this contract users can swap between pegged source and target tokens.
+ * @dev Make sure only to deploy with target ERC20 token having a fixed supply
+ * as this contract assumes a simple 1:1 fixed peg.
+ */
+contract ERC20TokenPeg is ISwap, Initializable {
+  using SafeMath for uint256;
+  using SafeERC20 for IERC20;
+
+  IERC20 private immutable _source;
+  IERC20 private immutable _target;
+
+  /**
+   * @dev Create the ERC20TokenPeg between two tokens
+   *
+   * Requirements:
+   * - `sourceAddr` and `targetAddr` cannot be the zero address.
+   * - `sourceAddr` and `targetAddr` cannot be the same contracts.
+   *
+   * @param sourceAddr The address of source token contract
+   * @param targetAddr The address of target token contract
+   */
+  constructor(
+    address sourceAddr,
+    address targetAddr
+  )
+    public
+  {
+    require(sourceAddr != address(0), "ERC20TokenPeg: sourceAddr is zero address");
+    require(targetAddr != address(0), "ERC20TokenPeg: targetAddr is zero address");
+    require(sourceAddr != targetAddr, "ERC20TokenPeg: sourceAddr == targetAddr");
+    _source = IERC20(sourceAddr);
+    _target = IERC20(targetAddr);
+  }
+
+  /**
+   * @dev Initialize the ERC20TokenPeg, only once!
+   *
+   * Contract will transfer from caller and lock source token in the amount of
+   * the circulating supply of target token. Additionally, the contract will
+   * transfer from caller and lock target tokens in the amount of the
+   * uncirculated supply (callers balance).
+   *
+   * Requirements:
+   * - the contract must have allowance for `caller`'s source tokens of at least
+   * target token circulating supply `amount`.
+   * - the contract must have allowance for `caller`'s target tokens of at least
+   * target token uncirculated supply `amount`.
+   */
+  function init()
+    external
+    initializer()
+  {
+    uint256 lockedSupply = _target.balanceOf(msg.sender);
+    uint256 circulatingSupply = _target.totalSupply().sub(lockedSupply);
+    // Transfer source tokens from sender to cover target's circulating supply
+    _source.safeTransferFrom(msg.sender, address(this), circulatingSupply);
+    // Transfer target tokens from sender to lock uncirculated supply
+    _target.safeTransferFrom(msg.sender, address(this), lockedSupply);
+  }
+
+  /// @inheritdoc ISwap
+  function source()
+    external
+    view
+    override
+    returns (address)
+  {
+    return address(_source);
+  }
+
+  /// @inheritdoc ISwap
+  function target()
+    external
+    view
+    override
+    returns (address)
+  {
+    return address(_target);
+  }
+
+  /**
+   * @dev the contract must have allowance for `caller`'s source tokens of at least `amount`.
+   * @inheritdoc ISwap
+   */
+  function deposit(uint256 amount)
```

Comment:
imo deposit and withdraw are overloaded / confusing and should be avoided, i would prefer just one method for swapPeg(source, destination) or a similar name.

Reply:
We could use one function to perform both `deposit` and `withdraw`, with an enum parameter representing the direction. Maybe something like:

```javascript
enum TokenOut { Source, Target }

...

function swapTo(TokenOut tokenOut, uint256 amount) { ...
```

Reply:
I used wETH9 as an inspiration, as users are familiar with that mechanism & interface. There is also the new and improved [wETH10](https://github.com/WETH10/WETH10/blob/main/contracts/interfaces/IWETH10.sol).

Then I tried to use the same interface for both wrapper and peg contracts, as they are trying to achieve the same thing using slightly different implementation. I think that the current interface nicely abstracts the desired concept and fits both contracts.

This interface is also easy to implement, while something like `swap(source, destination, amount)` is more for a general-purpose pool maybe even containing more than 2 tokens. And it does not fit the wrapper use-case where the destination is the contract you are calling. So now we could split interfaces and have two...

I did not want to go in that direction, so for this proposal I made a simple one that is IMO a good enough abstraction to fit both of these contracts. And my next suggestion would be to extend it with few more functions using the wETH10 interface as a template.

Reply:
The difference between this and wETH is when you're depositing there, you're using ETH to receive a token. Then with withdrawing you're supplying a token to receive ETH. Here these same terms are confusing because we're still dealing with token-to-token swaps.

I'm good with breaking the interface between the wrapper and peg contracts to make it more clear to users which asset they should supply and what they will receive.

Reply:
Yes, could we please move to swap(source, destination, amount), i have a preference for it